PYIC-8791: greater logging granularity on correlation failure types

thebauSoftwire opened the pull request:
## Proposed changes
### What changed

- We now have logs for when a user fails:
  - name correlation only
  - dob correlation only 
  - name AND dob correlation
- Updated unit test as it was checking testing for more than one thing (name failure and DOB failure) and was only passing because it had a name failure but in reality, we expected it to throw as the VC had a missing DOB despite being an identity credential
- added a reset function to the LoggerCollector to clear all log messages. This is so that a shared log collector can be re-used between tests.

### Why did it change

- KIWI are investigating a new feature and wanted to observe the number of users facing different types of correlation failures in order to inform their decision-making. Previously, the correlation checks were performed one after the other so only the first failure type was logged. They want visibility on the number of users who fail both name and DOB correlation so we need to check for both name and dob correlation and log the failure type appropriately.

### Issue tracking
<!-- Jira ticket & other docs, like RFCs -->

- [PYIC-8791](https://govukverify.atlassian.net/browse/PYIC-8791)

## Checklists

- [ ] READMEs and documentation up-to-date
- [x] API/ unit/ contract tests have been written/ updated
- [ ] No risk of exposure: PII, credentials, etc through logs/ code
- [ ] Production changes appropriately staged out


[PYIC-8791]: https://govukverify.atlassian.net/browse/PYIC-8791?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

thebauSoftwire committed:
PYIC-8791: update areVcsCorrelated to show more granularity in correlation failures

Previously, correlation failures were checked successively which means only the first failure is logged and we can't determine if users fail both name and DOB correlation. KIWI requested this update so they can better inform their decision-making by seeing the volume of correlation failure types users experience

thebauSoftwire committed:
PYIC-8791: update unit tests

One of the unit tests was only passing because the tests were being done successively and tested for more than one thing. I've updated the unit test and added a new one to cover a missing dob when it's required

thebauSoftwire commented on libs/user-identity-service/src/test/java/uk/gov/di/ipv/core/library/useridentity/service/UserIdentityServiceTest.java:
```diff
@@ -294,13 +294,34 @@ void areVCsCorrelatedReturnFalseWhenNameDifferentForBavCRI() throws Exception {
                                 createCredentialWithNameAndBirthDate(
-                                        "Jimmy", "Jones",
-                                        ""))); // BAV cri doesn't provide birthdate
+                                        "Jimmy", "Jones", "1000-01-01")));
```
This test was passing before since it was checking name correlation first and returned false when it had failed. Is there a reason we want to prioritise name correlation failures over a failed dob check? I'd expect this to throw since the VC would be invalid since it's an identity check credential with a missing DOB

DanCorderIPV replied:
Given we have a comment here saying that BAV doesn't provide DoBs should this be throwing an exception? Is the comment wrong, or are we OK with a missing DoB, or is it that the DoB field shouldn't exist at all in this case?

thebauSoftwire replied:
yeah I did wonder about that comment. I found the original PR which [ignores BAV for DOB correlation](https://github.com/govuk-one-login/ipv-core-back/pull/2736/changes#diff-03e309f6bbe7b1e939e8d7a3f5ad31be61b375cf9c88a8a05048bed7970df470) but [this PR](https://github.com/govuk-one-login/ipv-core-back/pull/2736/changes) much later on that removes it from the exclusion list so I assumed it was an old test 🤔 @blakeyp it looks like you put up the latter. Do you know if we should be excluding the BAV VCs for DOB correlation?

blakeyp replied:
Ah so the original HMRC-backed BAV CRI didn’t give us DOB, but the Experian-backed version does (which is what actually went live). So the comment/test case is out of date, we should be checking it

DanCorderIPV commented on libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/useridentity/service/UserIdentityService.java:
```diff
@@ -169,16 +169,20 @@ public boolean areVcsCorrelated(List<VerifiableCredential> vcs)
             LOGGER.error(LogHelper.buildErrorMessage(ErrorResponse.FAILED_BIRTHDATE_CORRELATION));
-            return false;
+        } else if (!successfulNameCorrelation) {
+            LOGGER.error(LogHelper.buildErrorMessage(ErrorResponse.FAILED_NAME_CORRELATION));
```
Given that someone cared enough about to this to raise the ticket we should probably add tests that check the error logging (or add to existing tests)

thebauSoftwire committed:
PYIC-8791: add tests for checking logs from failed correlations

DanCorderIPV commented on libs/user-identity-service/src/test/java/uk/gov/di/ipv/core/library/useridentity/service/UserIdentityServiceTest.java:
```diff
@@ -558,7 +598,13 @@ void areVCsCorrelatedShouldReturnFalseIfBavHasDifferentBirthDate() throws Except
         assertFalse(userIdentityService.areVcsCorrelated(vcs));
+
+        var logMessages = logCollector.getLogMessages();
+        //        assertEquals(1, logMessages.size());
```
uncomment?

DanCorderIPV commented on libs/user-identity-service/src/test/java/uk/gov/di/ipv/core/library/useridentity/service/UserIdentityServiceTest.java:
```diff
@@ -294,11 +302,64 @@ void areVCsCorrelatedReturnFalseWhenNameDifferentForBavCRI() throws Exception {
+                                BAV,
+                                createCredentialWithNameAndBirthDate(
+                                        "Jimbo", "Jones",
+                                        ""))); // BAV cri doesn't provide birthdate
```
I think we need to change/drop this comment

thebauSoftwire committed:
PYIC-8791: use a shared LogCollector and clear the log messages collected before running each test

adding appenders with the same name to the Logger for a given class I think was breaking the tests - the returned appender from getLogCollectorFor doesn't receive any log messages possibly due to the conflicting appender names

DanCorderIPV reviewed (approved):
Thanks for tidying up the tests too
